File: app-not-subscribed-to-event/node/github-app-event-subscriptions.mjs
/**
 * Find webhook events a GitHub App's handlers wait for and never receive.
 *
 * Read only. Two GETs with the App's JWT: the App's own record and its recent
 * webhook deliveries. Subscribing is an edit to the App and then a human
 * acceptance on every installation, so the script prints the three steps in
 * the order they have to happen.
 *
 * This is the App-side case. A repository or organization webhook has its own
 * events list, which is a different object with a different repair.
 *
 * Environment:
 *   GITHUB_APP_JWT   the JWT your own signing code produced
 */
const API = 'https://api.github.com';
const UA = 'github-app-event-subscriptions/1.0';

/** Which App permission gates which event. Curated, not fetched. */
export const EVENT_PERMISSION = {
  check_run: 'checks',
  check_suite: 'checks',
  commit_comment: 'contents',
  create: 'contents',
  delete: 'contents',
  deployment: 'deployments',
  deployment_status: 'deployments',
  fork: 'metadata',
  issue_comment: 'issues',
  issues: 'issues',
  label: 'metadata',
  member: 'members',
  membership: 'members',
  milestone: 'issues',
  organization: 'members',
  public: 'metadata',
  pull_request: 'pull_requests',
  pull_request_review: 'pull_requests',
  pull_request_review_comment: 'pull_requests',
  pull_request_review_thread: 'pull_requests',
  push: 'contents',
  release: 'contents',
  repository: 'metadata',
  repository_dispatch: 'contents',
  star: 'metadata',
  status: 'statuses',
  team_add: 'members',
  watch: 'metadata',
  workflow_dispatch: 'actions',
  workflow_job: 'actions',
  workflow_run: 'actions',
};

/** Permissions every App holds implicitly. */
export const ALWAYS_HELD = ['metadata'];

/** An event name reduced to the form GitHub spells it in. Pure. */
export function normalize(event) {
  return String(event ?? '').trim().toLowerCase();
}

/** The App permission that gates an event, or null if unknown. Pure. */
export function gatingPermission(event) {
  return EVENT_PERMISSION[normalize(event)] ?? null;
}

/** Whether the App holds a permission at read or better. Pure. */
export function holds(permissions, name) {
  if (ALWAYS_HELD.includes(name)) return true;
  const value = (permissions ?? {})[name];
  return Boolean(value) && String(value).trim().toLowerCase() !== 'none';
}

/** Distinct event names in a delivery log page. Pure. */
export function seenEvents(deliveries) {
  const out = new Set();
  for (const row of deliveries ?? []) {
    if (row && typeof row === 'object' && row.event) out.add(normalize(row.event));
  }
  return out;
}

/** Sort one handled event into a state. Pure. */
export function subscriptionState(event, subscribed, permissions, seen = null) {
  const name = normalize(event);
  const declared = new Set((subscribed ?? []).map(normalize));
  const gate = gatingPermission(name);
  if (declared.has(name)) {
    if (seen && seen.has(name)) {
      return ['subscribed-and-arriving',
        `${name} is declared by the App and has arrived recently.`];
    }
    return ['subscribed-not-yet-seen',
      `${name} is declared by the App but has not arrived in the retention ` +
      'window, which usually means it has not happened rather than that it ' +
      'is broken.'];
  }
  if (gate === null) {
    return ['not-subscribed-gate-unknown',
      `${name} is not declared by the App, so it has never been delivered. ` +
      'This script does not know which permission gates it; check the ' +
      'published event list before requesting one.'];
  }
  if (!holds(permissions, gate)) {
    return ['not-subscribed-blocked',
      `${name} is not declared, and the ${gate} permission that gates it is ` +
      'not held. The subscription cannot be ticked until the permission is added.'];
  }
  return ['not-subscribed-permitted',
    `${name} is not declared, but the ${gate} permission that gates it is ` +
    'held, so subscribing is an edit to the App followed by an acceptance round.'];
}

/** One row per handled event, in the order they were given. Pure. */
export function rows(handled, subscribed, permissions, seen = null) {
  return (handled ?? []).map((event) => {
    const [state, detail] = subscriptionState(event, subscribed, permissions, seen);
    return { event: normalize(event), state, detail, gated_by: gatingPermission(event) };
  });
}

/** Turn the rows into one finding. Pure. */
export function verdict(report) {
  const all = report ?? [];
  if (!all.length) {
    return ['nothing-handled',
      'no handled events were supplied, so there is nothing to compare the ' +
      "App's subscriptions against."];
  }
  const unreachable = all.filter((r) => r.state.startsWith('not-subscribed'));
  if (unreachable.length) {
    return ['handlers-unreachable',
      `${unreachable.length} of ${all.length} handled event(s) can never ` +
      'fire, because the App does not declare them.'];
  }
  const quiet = all.filter((r) => r.state === 'subscribed-not-yet-seen');
  if (quiet.length) {
    return ['all-subscribed-some-quiet',
      `every handled event is declared. ${quiet.length} of them has not ` +
      'arrived in the retention window, which is not by itself a fault.'];
  }
  return ['all-subscribed',
    'every handled event is declared by the App and arriving.'];
}

/** The ordered repair, as lines. Pure. */
export function repairSteps(report) {
  const all = report ?? [];
  const blocked = [...new Set(all
    .filter((r) => r.state === 'not-subscribed-blocked' && r.gated_by)
    .map((r) => r.gated_by))].sort();
  const missing = [...new Set(all
    .filter((r) => r.state.startsWith('not-subscribed'))
    .map((r) => r.event))].sort();
  if (!missing.length) return [];
  const steps = [];
  if (blocked.length) {
    steps.push(`add the ${blocked.join(', ')} permission to the App; until ` +
      'then the subscription cannot be selected at all');
  }
  steps.push(`subscribe the App to ${missing.join(', ')}`);
  steps.push('have an owner on every installation accept the resulting ' +
    'permission request, or the event will arrive from some accounts and not others');
  return steps;
}

async function get(jwt, path) {
  const url = path.startsWith('/') ? API + path : path;
  const res = await fetch(url, {
    headers: {
      Authorization: `Bearer ${jwt}`,
      Accept: 'application/vnd.github+json',
      'X-GitHub-Api-Version': '2022-11-28',
      'User-Agent': UA,
    },
  });
  let body = null;
  try { body = await res.json(); } catch { body = null; }
  return { status: res.status, body };
}

async function main() {
  const jwt = (process.env.GITHUB_APP_JWT || "dummy-github-app-jwt");
  if (!jwt) {
    console.error('set GITHUB_APP_JWT to the JWT your own signing code ' +
      "produced. The App's events array is on the App record, which an " +
      'installation token cannot read');
    process.exitCode = 2;
    return;
  }
  const handled = (process.argv[2] ?? '').split(',').map((s) => s.trim()).filter(Boolean);
  if (!handled.length) {
    console.error('pass the event names your receiver implements as the first ' +
      'argument, comma separated; without them there is nothing to compare');
    process.exitCode = 2;
    return;
  }

  const app = await get(jwt, '/app');
  if (app.status !== 200 || !app.body || typeof app.body !== 'object') {
    console.error(`GET /app returned ${app.status}, so the App's ` +
      'subscriptions cannot be read');
    process.exitCode = 2;
    return;
  }
  const subscribed = app.body.events ?? [];
  const permissions = app.body.permissions ?? {};
  console.log(`app subscribes to ${subscribed.length} event(s), holds ` +
    `${Object.keys(permissions).length} permission(s)`);

  let seen = null;
  const log = await get(jwt, '/app/hook/deliveries?per_page=100');
  if (log.status === 200 && Array.isArray(log.body)) {
    seen = seenEvents(log.body);
    console.log(`delivery log shows ${seen.size} distinct event(s) in the ` +
      'retention window');
  } else {
    console.log(`delivery log unavailable (${log.status}); the subscription ` +
      'answer does not depend on it');
  }

  const report = rows(handled, subscribed, permissions, seen);
  const [state, detail] = verdict(report);
  console.log(`${state}: ${detail}`);
  for (const row of report) {
    if (row.state !== 'subscribed-and-arriving') console.log(`  ${row.detail}`);
  }
  repairSteps(report).forEach((line, i) => {
    console.log(`repair step ${i + 1}: ${line}`);
  });

  console.log(JSON.stringify({
    subscribed: subscribed.map(normalize).sort(),
    permissions,
    seen_in_deliveries: seen ? [...seen].sort() : null,
    state,
    events: report,
  }, null, 2));
  process.exitCode = state === 'handlers-unreachable' ? 1 : 0;
}

// Only run when invoked directly, so importing this module from the test file
// does not execute main(), fail on the missing JWT and set an exit code that
// fails the suite even as every test passes.
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => { console.error(err.message); process.exitCode = 2; });
}

File: app-not-subscribed-to-event/python/github_app_event_subscriptions.py
"""Find webhook events a GitHub App's handlers wait for and never receive.

Read only. Two GETs with the App's JWT: the App's own record and its recent
webhook deliveries. Nothing is subscribed, permitted or changed. Subscribing is
an edit to the App and then a human acceptance on every installation, so the
script prints the three steps in the order they have to happen.

A GitHub App receives only the events it declares, and it can only declare an
event whose gating permission it holds. When the permission is absent the
subscription checkbox is not offered at all, so the subscription silently never
happens and nothing anywhere records the attempt.

This is the App-side case. A repository or organization webhook has its own
events list, read from the repository hooks endpoint, which is a different
object with a different repair.

Environment:

    GITHUB_APP_JWT   the JWT your own signing code produced
"""
import argparse
import json
import logging
import os
import sys

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("github_app_event_subscriptions")

API = "https://api.github.com"
UA = "github-app-event-subscriptions/1.0"

# Which App permission gates which event. Nothing in the API returns this, so
# the table is written from the documentation rather than fetched, and an event
# that is not in it is reported as unknown rather than guessed at. A wrong
# answer here sends somebody to request a permission they do not need.
EVENT_PERMISSION = {
    "check_run": "checks",
    "check_suite": "checks",
    "commit_comment": "contents",
    "create": "contents",
    "delete": "contents",
    "deployment": "deployments",
    "deployment_status": "deployments",
    "fork": "metadata",
    "issue_comment": "issues",
    "issues": "issues",
    "label": "metadata",
    "member": "members",
    "membership": "members",
    "milestone": "issues",
    "organization": "members",
    "public": "metadata",
    "pull_request": "pull_requests",
    "pull_request_review": "pull_requests",
    "pull_request_review_comment": "pull_requests",
    "pull_request_review_thread": "pull_requests",
    "push": "contents",
    "release": "contents",
    "repository": "metadata",
    "repository_dispatch": "contents",
    "star": "metadata",
    "status": "statuses",
    "team_add": "members",
    "watch": "metadata",
    "workflow_dispatch": "actions",
    "workflow_job": "actions",
    "workflow_run": "actions",
}

# Every event carries metadata implicitly, so an App that holds nothing still
# holds this one. Listing it keeps the "not permitted" branch honest.
ALWAYS_HELD = ("metadata",)


def normalize(event):
    """An event name reduced to the form GitHub spells it in. Pure.

    Case and surrounding whitespace only. A genuinely misspelled name is left
    misspelled so it stays visible as unknown, because quietly correcting it
    would hide the actual mistake in a report about missing events.
    """
    return str(event or "").strip().lower()


def gating_permission(event):
    """The App permission that gates an event, or None if unknown. Pure."""
    return EVENT_PERMISSION.get(normalize(event))


def holds(permissions, name):
    """Whether the App holds a permission at read or better. Pure."""
    if name in ALWAYS_HELD:
        return True
    value = (permissions or {}).get(name)
    return bool(value) and str(value).strip().lower() != "none"


def seen_events(deliveries):
    """Distinct event names in a delivery log page. Pure."""
    out = set()
    for row in deliveries or []:
        if isinstance(row, dict) and row.get("event"):
            out.add(normalize(row["event"]))
    return out


def subscription_state(event, subscribed, permissions, seen=None):
    """Sort one handled event into a state. Pure.

    subscribed is the App's events array, permissions its permission map, seen
    the distinct events observed in the delivery log. seen only ever refines a
    positive answer: an event that has not arrived may simply not have
    happened, so its absence is never a finding on its own.
    """
    name = normalize(event)
    declared = {normalize(e) for e in (subscribed or [])}
    gate = gating_permission(name)
    if name in declared:
        if seen is not None and name in seen:
            return ("subscribed-and-arriving",
                    "%s is declared by the App and has arrived recently." % name)
        return ("subscribed-not-yet-seen",
                "%s is declared by the App but has not arrived in the "
                "retention window, which usually means it has not happened "
                "rather than that it is broken." % name)
    if gate is None:
        return ("not-subscribed-gate-unknown",
                "%s is not declared by the App, so it has never been "
                "delivered. This script does not know which permission gates "
                "it; check the published event list before requesting one."
                % name)
    if not holds(permissions, gate):
        return ("not-subscribed-blocked",
                "%s is not declared, and the %s permission that gates it is "
                "not held. The subscription cannot be ticked until the "
                "permission is added." % (name, gate))
    return ("not-subscribed-permitted",
            "%s is not declared, but the %s permission that gates it is held, "
            "so subscribing is an edit to the App followed by an acceptance "
            "round." % (name, gate))


def rows(handled, subscribed, permissions, seen=None):
    """One row per handled event, in the order they were given. Pure."""
    out = []
    for event in handled or []:
        state, detail = subscription_state(event, subscribed, permissions, seen)
        out.append({"event": normalize(event), "state": state, "detail": detail,
                    "gated_by": gating_permission(event)})
    return out


def verdict(report):
    """Turn the rows into one finding. Pure."""
    report = report or []
    if not report:
        return ("nothing-handled",
                "no handled events were supplied, so there is nothing to "
                "compare the App's subscriptions against.")
    unreachable = [r for r in report if r["state"].startswith("not-subscribed")]
    if unreachable:
        return ("handlers-unreachable",
                "%d of %d handled event(s) can never fire, because the App "
                "does not declare them." % (len(unreachable), len(report)))
    quiet = [r for r in report if r["state"] == "subscribed-not-yet-seen"]
    if quiet:
        return ("all-subscribed-some-quiet",
                "every handled event is declared. %d of them has not arrived "
                "in the retention window, which is not by itself a fault."
                % len(quiet))
    return ("all-subscribed",
            "every handled event is declared by the App and arriving.")


def repair_steps(report):
    """The ordered repair, as lines. Pure.

    Order matters and is the point of the note: add the permission, subscribe,
    then have installations accept. Doing only the last is the usual mistake.
    """
    blocked = sorted({r["gated_by"] for r in report or []
                      if r["state"] == "not-subscribed-blocked" and r["gated_by"]})
    missing = sorted({r["event"] for r in report or []
                      if r["state"].startswith("not-subscribed")})
    if not missing:
        return []
    steps = []
    if blocked:
        steps.append("add the %s permission to the App; until then the "
                     "subscription cannot be selected at all"
                     % ", ".join(blocked))
    steps.append("subscribe the App to %s" % ", ".join(missing))
    steps.append("have an owner on every installation accept the resulting "
                 "permission request, or the event will arrive from some "
                 "accounts and not others")
    return steps


def get(session, path):
    """One GET. Returns (status, json-or-None)."""
    url = API + path if path.startswith("/") else path
    r = session.get(url, timeout=30)
    try:
        body = r.json()
    except ValueError:
        body = None
    return r.status_code, body


def main():
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--handles", default="",
                    help="comma-separated event names your handlers implement, "
                         "spelled the way GitHub spells them")
    args = ap.parse_args()

    jwt = os.environ.get("GITHUB_APP_JWT")
    if not jwt:
        log.error("set GITHUB_APP_JWT to the JWT your own signing code "
                  "produced. The App's events array is on the App record, "
                  "which an installation token cannot read")
        return 2

    handled = [h for h in (p.strip() for p in args.handles.split(",")) if h]
    if not handled:
        log.error("pass --handles with the event names your receiver "
                  "implements; without them there is nothing to compare")
        return 2

    session = requests.Session()
    session.headers.update({
        "Authorization": "Bearer " + jwt,
        "Accept": "application/vnd.github+json",
        "X-GitHub-Api-Version": "2022-11-28",
        "User-Agent": UA,
    })

    status, app = get(session, "/app")
    if status != 200 or not isinstance(app, dict):
        log.error("GET /app returned %d, so the App's subscriptions cannot be "
                  "read", status)
        return 2
    subscribed = app.get("events") or []
    permissions = app.get("permissions") or {}
    log.info("app subscribes to %d event(s), holds %d permission(s)",
             len(subscribed), len(permissions))

    seen = None
    d_status, deliveries = get(session, "/app/hook/deliveries?per_page=100")
    if d_status == 200 and isinstance(deliveries, list):
        seen = seen_events(deliveries)
        log.info("delivery log shows %d distinct event(s) in the retention "
                 "window", len(seen))
    else:
        log.info("delivery log unavailable (%d); the subscription answer does "
                 "not depend on it", d_status)

    report = rows(handled, subscribed, permissions, seen)
    state, detail = verdict(report)
    log.info("%s: %s", state, detail)
    for row in report:
        if row["state"] != "subscribed-and-arriving":
            log.info("  %s", row["detail"])

    for i, line in enumerate(repair_steps(report), start=1):
        log.info("repair step %d: %s", i, line)

    print(json.dumps({"subscribed": sorted(normalize(e) for e in subscribed),
                      "permissions": permissions,
                      "seen_in_deliveries": sorted(seen) if seen is not None else None,
                      "state": state, "events": report}, indent=2, default=str))
    return 1 if state == "handlers-unreachable" else 0


if __name__ == "__main__":
    sys.exit(main())
